Source/Mothership/MPIHandlers.cpp

```diff
@@ -0,0 +1,497 @@
+    if (!decode_app_spec_message(message, &appName, &distCount,
+                                 &appNumber))
+    {
+        debug_post(591, 3, "Q::APP,Q::SPEC", hex2str(message->Key()).c_str(),
+                   "Failed to decode.");
+        return 0;
+    }
+
+    debug_post(591, 3, "Q::APP,Q::SPEC", hex2str(message->Key()).c_str(),
+               dformat("appName=%s, distCount=%u, appNumber=%u",
+                       appName.c_str(), distCount, appNumber).c_str());
+
+    /* Ensure application existence idempotently (it might have been created by
+     * an AppDist message). */
+    appInfo = appdb.check_create_app(appName, distCount);
+
+    /* If the application is not in the UNDERDEFINED state, post angrily and do
+     * nothing else. */
+    if(appInfo->state != UNDERDEFINED)
+    {
+        Post(502, appName, appName, appInfo->get_state_colloquial());
+        return 0;
+    }
+
+    /* Define the application number in the database. */
+    appdb.numberToApp[appNumber] = appName;
+
+    /* Otherwise, set the distCount. */
+    appInfo->distCountExpected = distCount;
+
+    /* Check for being fully defined (transition from UNDERDEFINED to
+     * DEFINED). */
+    appInfo->check_update_defined_state();
+
+    /* Check for further state transitions. */
+    if (appInfo->should_we_recall()) recall_application(appInfo);
+    else if (appInfo->should_we_continue()) initialise_application(appInfo);
+    return 0;
+}
+
+unsigned Mothership::handle_msg_app_dist(PMsg_p* message)
+{
+    AppInfo* appInfo;
+    CoreInfo* coreInfo;
+
+    /* Pull message contents. */
+    std::string appName;
+    std::string codePath;
+    std::string dataPath;
+    uint32_t coreAddr;
+    std::vector<uint32_t>::iterator threadAddrIt;
+    std::vector<uint32_t> threadsExpected;
+    if (!decode_app_dist_message(message, &appName, &codePath,
+                                 &dataPath, &coreAddr,
+                                 &threadsExpected))
+    {
+        debug_post(591, 3, "Q::APP,Q::DIST", hex2str(message->Key()).c_str(),
+                   "Failed to decode.");
+        return 0;
+    }
+
+    #if ORCHESTRATOR_DEBUG
+    std::string threadNames = "";
+    for (threadAddrIt = threadsExpected.begin();
+         threadAddrIt != threadsExpected.end(); threadAddrIt++)
+        threadNames.append(dformat("0x%x ", *threadAddrIt));
+    debug_post(591, 3, "Q::APP,Q::DIST", hex2str(message->Key()).c_str(),
+               dformat("appName=%s, codePath=%s, dataPath=%s, coreAddr=0x%x, "
+                       "threadsExpected=%s",
+                       appName.c_str(), codePath.c_str(), dataPath.c_str(),
+                       coreAddr, threadNames.c_str()).c_str());
+    #endif
+
+    /* Ensure application existence idempotently. */
+    appInfo = appdb.check_create_app(appName);
+
+    /* If the application is not in the UNDERDEFINED state, don't post angrily
+     * (otherwise we'll have hundreds of post messages). Given that an AppSpec
```

Comment:
Not for this case, because the extra messages are completely meaningless. Even the number of messages is not useful, because that could be checked from the dist counters.

In general I agree, but I don't feel we're at that point yet.